File: src/main/java/at/yawk/fiction/impl/fanfiction/ChapterPageParser.java
package at.yawk.fiction.impl.fanfiction;

import at.yawk.fiction.*;
import at.yawk.fiction.impl.PageParser;
import java.util.ArrayList;
import java.util.List;
import javax.annotation.Nullable;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

/**
 * @author yawkat
 */
class ChapterPageParser extends PageParser<StoryChapterPair> {
    @SuppressWarnings("unchecked")
    @Override
    protected StoryChapterPair parse(Element root) throws Exception {
        Element profileTop = root.select("#profile_top").first();

        if (profileTop == null) { throw new NotFoundException(); }

        Story.StoryBuilder storyBuilder = Story.builder();
        storyBuilder.title(profileTop.select("> b.xcontrast_txt").text());
        // todo: author, tags

        @Nullable Element select = root.select("#chap_select").first();
        int chapterCount = 1;
        Elements chapterOptions = null;
        if (select != null) {
            chapterOptions = select.select("option");
            chapterCount = chapterOptions.size();
        }
        List<Chapter> chapters = new ArrayList<>();
        storyBuilder.chapters(chapters);

        for (int i = 0; i < chapterCount; i++) {
            Chapter.ChapterBuilder chapterBuilder = Chapter.builder();
            chapterBuilder.metadata(new FfnChapterMetadata(i));
            if (select != null) {
                String name = chapterOptions.get(i).text();
                chapterBuilder.name(name.substring(name.indexOf(' ') + 1)); // remove leading enumeration
            }
            chapters.add(chapterBuilder.build());
        }

        Chapter.ChapterBuilder currentChapter = Chapter.builder();
        currentChapter.text(new HtmlText(root.select("#storytext").html()));

        return new StoryChapterPair(storyBuilder.build(), currentChapter.build());
    }
}

File: src/main/kotlin/at/yawk/fiction/impl/fanfiction/ChapterPageParser.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package at.yawk.fiction.impl.fanfiction

import at.yawk.fiction.Chapter
import at.yawk.fiction.HtmlText
import at.yawk.fiction.NotFoundException
import at.yawk.fiction.Story
import at.yawk.fiction.impl.PageParser
import org.jsoup.nodes.Element
import org.jsoup.select.Elements
import java.util.*

/**
 * @author yawkat
 */
internal object ChapterPageParser : PageParser<Pair<Story, Chapter>>() {
    override fun parse(root: Element): Pair<Story, Chapter> {
        val profileTop = root.select("#profile_top").first() ?: throw NotFoundException()

        val title = profileTop.select("> b.xcontrast_txt").text()
        // todo: author, tags

        val select = root.select("#chap_select").first()
        var chapterCount = 1
        var chapterOptions: Elements? = null
        if (select != null) {
            chapterOptions = select.select("option")
            chapterCount = chapterOptions!!.size
        }

        val chapters = ArrayList<Chapter>()
        for (i in 0..chapterCount - 1) {
            var name: String? = null
            if (chapterOptions != null) {
                val text = chapterOptions[i].text()
                name = text.substring(text.indexOf(' ') + 1)
            }
            chapters.add(Chapter(
                    name = name,
                    providerData = FfnChapterProviderData(index = i)
            ))
        }

        val text = HtmlText(root.select("#storytext").html())
        return Pair(Story(
                title = title,
                chapters = chapters,
                providerData = FfnStoryProviderData()
        ), Chapter(
                text = text,
                providerData = FfnChapterProviderData()
        ))
    }
}
